website/hw/hw4.rst

```diff
@@ -180,103 +218,517 @@ As you are stepping through your program, you might want to be able to
+  then print the value of variable ``a`` after line 41.
+- **Practice Question D**: What is the return value of the function ``num()`` in hexadecimal?
+
+You can find the answers to these questions at the end of this homework.
+
+The Micro Editor
+----------------
+
+We've seen how GDB works with some simple examples, but now it's time to debug
+a more complex piece of software. We will be using a very simple terminal-based
+editor called ``micro``. This is our version of the ``kilo`` editor, a simple
+but functional text editor that can be implemented in about 1,000 lines of C
+code; if you're interested, you can see a step-by-step guide to writing this
+editor here: https://viewsourcecode.org/snaptoken/kilo/ (please note that you
+do not have to read this to complete this homework; however, if you're interested
+in the inner workings of a text editor, you may find that guide interesting)
+
+``micro`` largely follows the same structure as the ``kilo`` code, except we
+divided it into multiple modules and documented the code following our style guide.
+
+You can build the ``micro`` editor by going into the ``micro/`` directory in
+your repository and running the following::
+
+    cmake -B build/
+    make -C build/
+
+This will generate a ``micro`` executable inside a ``build/`` directory.
+Let's give it a try! Run the following from the ``micro/`` directory::
+
+    build/micro
+
+This will open the editor with a "blank file". You can start typing to edit
+the file, and you'll notice that you can move around with the arrow keys, use
+the Backspace key, and the Delete key. You should be able to quit the editor
+by pressing Ctrl-Q (if you modified the file, you'll have to press it three
+times to confirm you want to exit without saving).
+
+However, we have inserted a few bugs in the code, so don't be surprised if
+you do something that makes the editor crash! Please note that, if this
+happens, your terminal may be left in a seemingly unusable state (in particular,
+it will seem like you can't type anything into the terminal). When this happens,
+press Enter followed by typing the word ``reset`` followed by pressing Enter (this doesn't
+reset your computer, it just resets your terminal). You can also just close the
+terminal and open it up again.
+
+Let's say we wanted to debug this executable. We actually cannot call GDB like
+before, because the text editor needs to use the entire terminal, so it
+would be impractical to have a GDB prompt interfering with that. So, this will
+allow us to do *remote debugging*, where the program and the debugger are
+run separately (and could even be running in completely different computers,
+and communicating over the network).
+
+To do this, open two terminals (if you're logging to a CS Linux server, make
+sure you're logged into the same Linux server on both terminals). On the first one, run the following from the
+``micro/`` directory::
```

Comment:
We may want to note that they might want 3 terminal windows open; one for the editor, one for gdb, and one to look at files without closing everything

Reply:
Done.